implementation/sequicity/tsd_net.py

```diff
@@ -291,22 +385,36 @@ def get_sparse_selective_input(self, x_input_np):
         return result
 
     def forward(self, z_enc_out, u_enc_out, u_input_np, m_t_input, degree_input, last_hidden, z_input_np):
+        """
+        This is an implementation of the decoder part of CopyNet. We want to
+        replace this with a Transformer Decoder with a Copy Mechanism.
+
+        Does:
+            1) Transformation of input into sparse - TODO: Look into this, might be important
+            2) Attention, GRU, generation/copy score = CopyNet - No need to understand closely, should be replaced.
+        """
         sparse_z_input = Variable(self.get_sparse_selective_input(z_input_np), requires_grad=False)
 
         m_embed = self.emb(m_t_input)
-        
-        # z_enc_out are stacked (along the time-step axis) outputs of the BSpan decoder
+
+        # z_enc_out are stacked (along the time-step axis) belief spans (computed in BSpanDecoder)
+        # TODO: Why should this be a BSpanDecoder output, when the variable states:
+        #   - z_ = belief span
+        #   - enc_out = encoder output
+        #   - Would a better name be z_dec_out, or is this incorrect understanding of the code?
```

Comment:
This is weird. See  call of `ResponseDecoder` (named `self.m_decoder`) in the main training loop of `forward_turn` function: 
`last_hidden, dec_out = self.m_decoder(pz_dec_outs, u_enc_out, u_input_np, m_tm1, degree_input, last_hidden, z_input_np)`
`pz_dec_outs` are stacked outputs in bspandecoder.

Reply:
Yeah, I agree. It is an error in their naming. `z_enc_out` should be -> `z_dec_out`.
Actually, is it always `u_enc_*` and `z_dec_*` and `m_dec_*`? Maybe all `z_enc_*` are wrong?